### DISCUSSION TITLE
smart-contract-verifier microservice not support Verify & publish contract by standard json input mode

### DISCUSSION THREAD

**Discussion by maxufeng** [blockscout/blockscout#10810] [Q&A] [answered]
The smart-contract-verifier service currently finds that it does not support standard json input mode to verify contracts and reports illegal json data. Through https://eth-bytecode-db.services.blockscout.com/ this service is to support the standard json input model validation of the contract. Is this a bug? Can we solve this problem later?

**Comment by GimluCom** [comment]
Could you provide some more details of the issue? An error message and any other information would help us to better understand the problem.

**Reply by maxufeng** [reply]
1. precondition

(1) contract code
`// SPDX-License-Identifier: GPL-3.0

pragma solidity 0.8.9;

/**
 * @title Storage
 * @dev Store & retrieve value in a variable
 * @custom:dev-run-script ./scripts/deploy_with_ethers.ts
 */
contract Storage {

    uint256 number;

    /**
     * @dev Store value in variable
     * @param num value to store
     */
    function store(uint256 num) public {
        number = num;
    }

    /**
     * @dev Return value 
     * @return value of 'number'
     */
    function retrieve() public view returns (uint256){
        return number;
    }
}
`
(2) json file 
![图片](https://github.com/user-attachments/assets/b3e4596a-ac1d-4222-989a-8c812ba16efe)

170781df9fb5f71f052a622634264fd7.json
[170781df9fb5f71f052a622634264fd7.json](https://github.com/user-attachments/files/17162043/170781df9fb5f71f052a622634264fd7.json)

2. execute with http://eth-bytecode-db.services.blockscout.com
Input - using official provided by http://eth-bytecode-db.services.blockscout.com
envs/common-blockscout.env config
 
![图片](https://github.com/user-attachments/assets/ec647a48-0dd0-414b-b538-e6692a829acc)

![图片](https://github.com/user-attachments/assets/bed524f3-5403-4b73-b32b-3429c51d44b8)
![图片](https://github.com/user-attachments/assets/6d3de756-f938-4c88-96c8-a9dd7b93f6b1)
（1）Expected results:
        Contract verification successful
（2）Actual results:
        content is not a valid standard json: missing field `language` at line 2798 column 1
（3）blockscout log
{"time":"2024-09-27T09:03:57.210Z","severity":"info","message":"Smart-contract verification via standard json input started","metadata":{}}
{"time":"2024-09-27T09:03:57.211Z","severity":"info","message":"Sent 200 in 4ms on POST /api/v2/smart-contracts/0xc643e683b6bfeca7dfbca75871ddd1c7e96ff74c/verification/via/standard-input?","metadata":{"status":"200","unit":"microsecond","request_id":"F_kOdzOx9MDd-LAAGuQh","method":"POST","duration":4059,"endpoint":"/api/v2/smart-contracts/0xc643e683b6bfeca7dfbca75871ddd1c7e96ff74c/verification/via/standard-input?"}}
{"time":"2024-09-27T09:03:57.721Z","severity":"error","message":"Smart-contract verification error: content is not a valid standard json: missing field `language` at line 2798 column 1","metadata":{}}
{"time":"2024-09-27T09:03:57.721Z","severity":"error","message":"Solidity smart-contract verification 0xc643e683b6bfeca7dfbca75871ddd1c7e96ff74c failed because of the error \"content is not a valid standard json: missing field `language` at line 2798 column 1\"","metadata":{}}
{"time":"2024-09-27T09:03:57.721Z","severity":"error","message":"Solidity smart-contract verification 0xc643e683b6bfeca7dfbca75871ddd1c7e96ff74c failed because of the error: #Ecto.Changeset<action: :insert, changes: %{name: \"undefined\", compiler_version: \"v0.8.9+commit.e5eed63a\", address_hash: %Explorer.Chain.Hash{byte_count: 20, bytes: <<198, 67, 230, 131, 182, 191, 236, 167, 223, 188, 167, 88, 113, 221, 209, 199, 233, 111, 247, 76>>}, partially_verified: false, autodetect_constructor_args: false, contract_code_md5: \"76b51db0a3ec87648b44da1a9c405883\", is_blueprint: false, is_vyper_contract: false, license_type: :gnu_gpl_v3, verified_via_eth_bytecode_db: false, verified_via_sourcify: false, verified_via_verifier_alliance: false}, errors: [files: {\"content is not a valid standard json: missing field `language` at line 2798 column 1\", []}], data: #Explorer.Chain.SmartContract<>, valid?: false>","metadata":{}}
{"time":"2024-09-27T09:03:57.721Z","severity":"info","message":"Smart-contract 0xc643e683b6bfeca7dfbca75871ddd1c7e96ff74c verification: broadcast verification results","metadata":{}}
{"time":"2024-09-27T09:03:57.721Z","severity":"info","message":"Broadcast smart-contract 0xc643e683b6bfeca7dfbca75871ddd1c7e96ff74c verification results","metadata":{}}


3. execute with smart-contract-verifier
envs/common-blockscout.env config
![图片](https://github.com/user-attachments/assets/1f6a7322-48cc-45f0-9b5f-ab597b49b732)


（1）Expected results:
        Contract verification successful
（2）Actual results:
        content is not a valid standard json: missing field `language` at line 2798 column 1
（3）smart-contract-verifier log
2024-09-27T09:15:32.466173Z  INFO HTTP request{method=POST endpoint=default client_ip=172.20.0.8 request_id=8f2c51dd-12c8-4325-9162-003e420378fa}: blockscout_service_launcher::launcher::span_builder: new
2024-09-27T09:15:32.466203Z  INFO HTTP request{method=POST endpoint=default client_ip=172.20.0.8 request_id=8f2c51dd-12c8-4325-9162-003e420378fa}: blockscout_tracing_actix_web::middleware: Started HTTP request processing
2024-09-27T09:15:32.467413Z  INFO HTTP request{method=POST endpoint=default client_ip=172.20.0.8 request_id=8f2c51dd-12c8-4325-9162-003e420378fa}: smart_contract_verifier_server::services::solidity_verifier: Solidity standard-json verification request received chain_id="111" contract_address="0xc643e683b6bfeca7dfbca75871ddd1c7e96ff74c"
2024-09-27T09:15:32.467700Z  INFO HTTP request{method=POST endpoint=default client_ip=172.20.0.8 request_id=8f2c51dd-12c8-4325-9162-003e420378fa}: smart_contract_verifier_server::services::solidity_verifier: Request processed response=VerifyResponse { message: "content is not a valid standard json: missing field `language` at line 2798 column 1", status: Failure, source: None, extra_data: None, post_action_responses: None }
2024-09-27T09:15:32.467737Z  INFO HTTP request{method=POST endpoint=default client_ip=172.20.0.8 request_id=8f2c51dd-12c8-4325-9162-003e420378fa duration=1532 unit="microsecond"}: blockscout_tracing_actix_web::middleware: Finished HTTP request processing
2024-09-27T09:15:32.467768Z  INFO HTTP request{method=POST endpoint=default client_ip=172.20.0.8 request_id=8f2c51dd-12c8-4325-9162-003e420378fa duration=1532 unit="microsecond"}: blockscout_service_launcher::launcher::span_builder: close time.busy=1.54ms time.idle=57.3µs



Looking forward to your reply

**Reply by GimluCom** [reply, marked as answer]
Please try this file [standard_input.json](https://github.com/user-attachments/files/17163200/standard_input.json)

**Reply by maxufeng** [reply]
The json file can be verified correctly. Just to see what standard Json input means, thank you

**Comment by maxufeng** [comment]
Developers can be compared with the problems 170781 df9fb5f71f052a622634264fd7. Json and standard_input json the difference between two files.

**Reply by GimluCom** [reply]
Here is also a visual example: [build-info.json](https://github.com/melisaerdal/Solidity/blob/ea38201678c5757d950de1b958d99452e60ed5c1/contracts...artifacts...build-info...dd0c335c286cb6b5735333c6c69940a3.json#L6C11-L40C3)
